---
title: "4_Pseudo-bulk_analysis_for_single-cell_RNA-Seq_data"
author: "Ji-Qing Chen"
date: "2024-06-29"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ExperimentHub)
library(Seurat)
library(DESeq2)
library(tidyverse)
```

# What is Pseudo-bulk analysis
```{r}
# Credit: https://www.youtube.com/watch?v=04gB2owLKus&list=PLJefJsd1yfhagnkss5B1YCsHaH0GWQfFT&index=8

# A certain group of cells that belong to a particular biological replicate or sample. We want to aggregate the counts and metadata for those cells belonging to a particular replicate or sample to the sample/replicate level. Finally, we have such a format where we have one count value for a gene for one replicate sample.
# Convert the single-cell feature barcode matrix into more sort of a bulk where we have the counts for aggregated for the sample or the replicate level and not at the single cell level.

# Then, leverage existing bulk RNA-seq DE frameworks, such as DESeq2, edgeR and limma.
```

# Why perform pseudo-bulk analysis?
```{r}
# Single-cell RNAseq data tend to exhibit an abundance of zero counts, a complicated distribution, and huge heterogeneity.

# The heterogeneity within and between cell population manifests major challenges to the differential gene expression analysis in scRNAseq data.

# Single-cell methods to identify highly expressed genes as DE and exhibit low sensitivity for genes having low expression.

# Single cell methods often inflate the p-values as each cell is treated as a sample.

# If cells are treated as samples, then variation across a population is not truly investigate. Because the cells are not truly independent of each other as they are isolated from the same organism or sample from the same environment. So they cannot be considered as samples because they are not truly independent of each other.

# If cells are treated as samples then the comparisons that we are making at the individual level between the cells and we are not truly looking at the variation across the population.

# pseudo-bulk approaches can dramatically reduce the number of zeros in the data especially for the genes that have low expression and at the same time we can also leverage the statistical rigor of existing bulk RNAseq differential expression (DE) methods like DESeq2, edgeR and limma.

# Also, DE testing performed on pseudo-bulk expression profiles leverages the resolution offered by single-cell technologies to define the labels.

# Each sample is presented no more then once for each condition, avoiding problems from unmodelled correlations between samples. It also takes care of situation where the cells are correlated to each other as they isolated from the same organism or the same sample so aggregating the counts at the sample level makes each sample be represented only once in a condition which can avoid the problems of correlation between the samples.

# To infer which genes might be important for a condition at the population level (not the individual level; investigate the effect of a condition or an effect of a treatment on genes), samples need to be acquired from different organisms/samples, not different cells.
```

# Get the data from the ExperimentHub
```{r}
eh <- ExperimentHub()
eh # this object will tell you what data it has from what provider, the species, the data class
query(eh, "Kang") # get information on all the data that is associated with "Kang"

# Get the data
sce <- eh[["EH2259"]] # accession ID
sce # It is single-cell experiment object. It need to be converted into Seurat object
seu.obj <- as.Seurat(sce, data = NULL) # It is the annotated data.
View(seu.obj@meta.data) # Annotation: meta.data$cell; also this data has been run on a doublet finder algorithm: meta.data$multiplets -> will filter out the doublets.
save(seu.obj, file = "/Users/chenjiqing/Public/scRNA_seq/Seurat_object_for_PseudoBulk_analysis.RData")
```

# QC and filtering
```{r}
# explore QC -> read previous script

# get mito percent
seu.obj$mitoPercent <- PercentageFeatureSet(seu.obj, pattern = '^MT-')
View(seu.obj@meta.data)

# filter low quality cells
seu.filtered <- subset(seu.obj, subset = nFeature_originalexp > 200 & nFeature_originalexp < 2500 &
         nCount_originalexp > 800 & 
         mitoPercent < 5 &
         multiplets == 'singlet')
# nFeature_originalexp: feature original experiment

seu.obj # started with around 29000 cells
seu.filtered # After filtering we are left with around 22000 cells
```

# Run Seurat's standard workflow steps
```{r}
seu.filtered <- NormalizeData(seu.filtered)
seu.filtered <- FindVariableFeatures(seu.filtered)
seu.filtered <- ScaleData(seu.filtered)
seu.filtered <- RunPCA(seu.filtered)
ElbowPlot(seu.filtered) # Determine the dimensionality of the data
seu.filtered <- RunUMAP(seu.filtered, dims = 1:20)

# visualize 
cell_plot <- DimPlot(seu.filtered, reduction = 'umap', group.by = 'cell', label = TRUE) # Group my cells as per the type of the cluster or type of the cells that forms the cluster.
cond_plot <- DimPlot(seu.filtered, reduction = 'umap', group.by = 'stim') # Group cells by condition.

cell_plot|cond_plot # cells are not integrated across conditions, and for this analysis it really does not matter whether my data is integrated or not because the data that I am accessing and I will be using is the raw counts. And even if you use an integrated data the method which returns the character expression matrix will use the normalized data from the data slot and not from the count slot which stores the raw counts. Hence, it really does not matter whether you have integrated data or not.

# In here, try to make comparisons in B cells between two conditions (control and stimulated) by aggregating the counts to the sample levels in these conditions and comparing between the samples and essentially identifying genes that have different expression patterns in the two condition groups.
```

# Pseudo-bulk workflow
```{r}
# Acquiring necessary metrics for aggregation across cells in a sample

# 1. counts matrix - sample level
# Aggregate the counts across the cells to the sample level.
View(seu.filtered@meta.data)
# "ind" column has information about the individuals that is the sample ID. We want to add the condition information to this as well. So I want to create new column with both sample ID and th condition information.
seu.filtered$samples <- paste0(seu.filtered$stim, seu.filtered$ind)

DefaultAssay(seu.filtered)

# Run a function to aggregate the counts across the cells.
cts <- AggregateExpression(seu.filtered, 
                    group.by = c("cell", "samples"),
                    assays = 'originalexp',
                    slot = "counts",
                    return.seurat = FALSE) # do not want to return seurat object
# group.by: we have to provide how to group the cells and aggregate. The first thing I want to group by is the cell type and for within each cell type I want to group by the individual or sample name. -> We want to aggregate by the cell types and the samples.
# assays: our default assay. use DefaultAssay to check.
# slot: the thing that you want to aggregate, in here, is counts because I will be using DESeq which uses raw counts.

cts <- as.matrix(cts$originalexp) # get the counts data by extracting it from the matrix
# column name: cell type + sample Name or ID; value is counts; rows are genes -> We want to aggregate by the cell types and the samples.
# First, it grouped the cells by the cell type, and then, it aggregated for each type of sample within it for each of the gene.
```

# Split the large matrix according to cell type (Goal: Because we want to compare the expression of genes in B cells across different conditions; we wanted our counts data to be aggregated at the sample level)
```{r}
# transpose
cts.t <- t(cts)

# convert to data.frame
cts.t <- as.data.frame(cts.t)
View(cts.t[1:10,1:10])

# get values where to split (get the cell type information from the rows; since it has the sample ID along with the condition I need to get rid of this pattern that is anything after the "_" and the values so I can get only the cell type information here)
splitRows <- gsub('_.*', '', rownames(cts.t))
# This can be used as a factor variable when we are using it to split the data frame.

# split data.frame (split rows according to the cell type into a list sort of an object where each element would correspond to a matrix associated with a cell type)
cts.split <- split.data.frame(cts.t,
                              f = factor(splitRows))
# f: define a grouping variable by providing the factor variable which contain all the cell types that we obtained from previous command.
# It returned a list with each element corresponding to a cell type and each of these elements corresponds to the matrix that is associated with the cell type.
# Having data in such a list object makes it convenient to fetch matrices for cell type in the data.

# Look at one of the element
View(cts.split$`B cells`[,1:10]) 
# a matrix that only corresponds to the aggregated values for B cells across all the samples and it has information for all the genes

# fix colnames and transpose (need to re-transpose the matrix so we need to make our rows as the gene names and the columns as the sample IDs; we also need to fix the column names: we need to get rid of the cell type names [ex. "B cell_"] from the column names)
cts.split.modified <- lapply(cts.split, function(x){
  rownames(x) <- gsub('.*_(.*)', '\\1', rownames(x))
  t(x)
})
# lapply: performs these operations on all the list objects and returns a list object as well.

View(cts.split.modified$`B cells`)
# That is specific to B cells and data aggregated to the sample level. 

# We can use these counts matrix into DESeq dataset; we can build DESeq dataset using the count matrix and we can run the DESeq analysis.
```

# Run DE analysis with B cells
```{r}
# 1. Get counts matrix
counts_bcell <- cts.split.modified$`B cells` # Having data in such a list object makes it convenient to fetch matrices for cell type in the data.
# resemble the bulk data

# 2. generate sample level metadata (since we do not have the metadata for the samples, we can create one)
colData <- data.frame(samples = colnames(counts_bcell))

colData <- colData %>%
  mutate(condition = ifelse(grepl('stim', samples), 'Stimulated', 'Control')) %>%
  column_to_rownames(var = 'samples')
# get more information from metadata (in here, condition status)
```

# Perform DESeq2
```{r}
# Create DESeq2 object   
dds <- DESeqDataSetFromMatrix(countData = counts_bcell,
                              colData = colData,
                              design = ~ condition)

# filter (remove all the genes that have lower than 10 reads)
keep <- rowSums(counts(dds)) >=10
dds <- dds[keep,]

# run DESeq2
dds <- DESeq(dds)

# Check the coefficients for the comparison
resultsNames(dds)

# Generate results object
res <- results(dds, name = "condition_Stimulated_vs_Control")
res
# Comparisons here were between the samples and not between cells.
```

